src/coreclr/src/jit/morph.cpp

```diff
@@ -18451,38 +18471,91 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
                     fieldSeq = nullptr;
                 }
             }
-            else if (indir->OperIs(GT_FIELD))
+        }
+        else
+        {
+            if (indir->OperIs(GT_IND))
             {
-                // TODO-ADDR: ObjectAllocator produces FIELD nodes with FirstElemPseudoField as field
-                // handle so we cannot use FieldSeqNode::GetFieldHandle() because it asserts on such
-                // handles. ObjectAllocator should be changed to create LCL_FLD nodes directly.
-                assert(fieldSeq->GetTail()->m_fieldHnd == indir->AsField()->gtFldHnd);
+                // Skip TYP_STRUCT IND nodes, it's not clear what we can do with them.
+                // Normally these should appear only as sources of variable sized copy block
+                // operations (DYN_BLK) so it probably doesn't make much sense to try to
+                // convert these to local nodes.
+                return;
             }
+
+            if ((user == nullptr) || !user->OperIs(GT_ASG))
+            {
+                // TODO-ADDR: Skip TYP_STRUCT indirs for now, unless they're used by an ASG.
+                // At least call args will require extra work because currently they must be
+                // wrapped in OBJ nodes so we can't replace those with local nodes.
+                return;
+            }
+
+            if (indir->OperIs(GT_FIELD))
+            {
+                CORINFO_CLASS_HANDLE fieldClassHandle;
+                CorInfoType          corType =
+                    m_compiler->info.compCompHnd->getFieldType(indir->AsField()->gtFldHnd, &fieldClassHandle);
+                assert(corType == CORINFO_TYPE_VALUECLASS);
+
+                structLayout = m_compiler->typGetObjLayout(fieldClassHandle);
+            }
+            else
+            {
+                structLayout = indir->AsBlk()->GetLayout();
+            }
+
+            // We're not going to produce a TYP_STRUCT LCL_FLD so we don't need the field sequence.
+            fieldSeq = nullptr;
         }
 
-        indir->ChangeOper(GT_LCL_FLD);
-        indir->AsLclFld()->SetLclNum(val.LclNum());
-        indir->AsLclFld()->SetLclOffs(val.Offset());
-        indir->AsLclFld()->SetFieldSeq(fieldSeq == nullptr ? FieldSeqStore::NotAField() : fieldSeq);
+        // We're only processing TYP_STRUCT variables now so the layout should never be null,
+        // otherwise the below layout equality check would be insufficient.
+        assert(varDsc->GetLayout() != nullptr);
+
+        if ((val.Offset() == 0) && (structLayout == varDsc->GetLayout()))
+        {
+            indir->ChangeOper(GT_LCL_VAR);
+            indir->AsLclVar()->SetLclNum(val.LclNum());
+        }
+        else if (!varTypeIsStruct(indir->TypeGet()))
+        {
+            indir->ChangeOper(GT_LCL_FLD);
```

Comment:
Does this block cover cases where we are lying about struct types, it means
`indir->TypeGet()` could be only int or long, correct?

Reply:
I believe that it could be byte or short as well as int or long.

Reply:
Well, the indir type could be any primitive type - `int`, `short`, `float` etc. You will get a `LCL_FLD` of that type, possibly having `NotAField` as field sequence if we're lying about types for whatever reason.

But I suppose you're referring to the specific case of struct return type from #1231. In that case the importer itself is producing a `LCL_FLD` and `LocalAddressVisitor` leaves it alone:
```
LocalAddressVisitor visiting statement:
STMT00000 (IL 0x000...0x005)
[000003] -AC-G-------              *  ASG       int   
[000001] U-----------              +--*  LCL_FLD   int    V08 loc0         [+0]
                                   +--*    ubyte  V08.a (offs=0x00) -> V12 tmp1         
                                   +--*    ubyte  V08.b (offs=0x01) -> V13 tmp2         
[000000] --C-G-------              \--*  CALL      int    Program.GetRetStruct

Local V08 should not be enregistered because: was accessed as a local field
```
Note that there's something fishy going in the particular example I have picked above: it's a 2 byte struct but the `LCL_FLD` produced by the importer has type `TYP_INT`. I don't know it doesn't have type `TYP_SHORT` but if a similar situation will arise with `IND` instead of `LCL_FLD`, `LocalAddressVisitor::EscapeLocation` will label the indir as "wide" (because it's large than the variable's real type), leave it alone and mark the variable "address exposed".

Reply:
I would guess that the JIT is retyping it because it always prefers to avoid small ints, and if the target is a local, it is safe. However, I wasn't easily able to figure out if and where that's being done in this case.